File: macos/Services/MCP/Tools/Tier3/InsertRowsTool.swift
// InsertRowsTool.swift
// Gridex
//
// MCP Tool: Insert rows into a table. Requires user approval.

import Foundation

struct InsertRowsTool: MCPTool {
    let name = "insert_rows"
    let description = "Insert one or more rows into a table. Requires user approval. Returns affected count."
    let tier = MCPPermissionTier.write

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "table_name": [
                "type": "string",
                "description": "Name of the table to insert into"
            ],
            "schema": [
                "type": "string",
                "description": "Optional schema name"
            ],
            "rows": [
                "type": "array",
                "description": "Array of row objects to insert",
                "items": ["type": "object"]
            ]
        ],
        "required": ["connection_id", "table_name", "rows"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let tableName = params["table_name"]?.stringValue else {
            throw MCPToolError.invalidParameters("table_name is required")
        }

        guard let rowsArray = params["rows"]?.arrayValue, !rowsArray.isEmpty else {
            throw MCPToolError.invalidParameters("rows must be a non-empty array")
        }

        // Check permission
        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        // Requires approval
        if permission.requiresUserApproval {
            let approved = await context.requestApproval(
                tool: name,
                description: "Insert \(rowsArray.count) row(s) into '\(tableName)'",
                details: formatRowsPreview(rowsArray),
                connectionId: connectionId
            )
            if !approved {
                throw MCPToolError.permissionDenied("User denied the operation.")
            }
        }

        let (adapter, _) = try await context.getAdapter(for: connectionId)
        let schemaName = params["schema"]?.stringValue

        var insertedCount = 0

        for rowValue in rowsArray {
            guard let rowObj = rowValue.objectValue else { continue }

            var values: [String: RowValue] = [:]
            for (key, val) in rowObj {
                values[key] = jsonValueToRowValue(val)
            }

            _ = try await adapter.insertRow(table: tableName, schema: schemaName, values: values)
            insertedCount += 1
        }

        return MCPToolResult(text: "Successfully inserted \(insertedCount) row(s) into '\(tableName)'.")
    }

    private func formatRowsPreview(_ rows: [JSONValue]) -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let previewRows = Array(rows.prefix(5))
        if let data = try? encoder.encode(previewRows),
           let str = String(data: data, encoding: .utf8) {
            if rows.count > 5 {
                return str + "\n... and \(rows.count - 5) more rows"
            }
            return str
        }
        return "\(rows.count) rows"
    }

    private func jsonValueToRowValue(_ value: JSONValue) -> RowValue {
        switch value {
        case .null: return .null
        case .bool(let b): return .boolean(b)
        case .int(let i): return .integer(Int64(i))
        case .double(let d): return .double(d)
        case .string(let s): return .string(s)
        case .array, .object:
            let encoder = JSONEncoder()
            if let data = try? encoder.encode(value),
               let str = String(data: data, encoding: .utf8) {
                return .json(str)
            }
            return .null
        }
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier3/InsertRowsTool.cpp
//
// InsertRowsTool.cpp
//

#include "InsertRowsTool.h"
#include "../MCPToolHelpers.h"
#include "../../Security/MCPIdentifierValidator.h"

namespace DBModels
{
    namespace
    {
        // Convert a JSON value to the wstring form expected by
        // TableRow. Nulls become the SQL-NULL sentinel so the
        // adapter emits `DEFAULT` / `NULL` rather than the literal
        // string "null".
        std::wstring jsonToCell(const nlohmann::json& v)
        {
            if (v.is_null())            return nullValue();
            if (v.is_boolean())         return v.get<bool>() ? L"true" : L"false";
            if (v.is_number_integer())  return std::to_wstring(v.get<int64_t>());
            if (v.is_number_float())    return std::to_wstring(v.get<double>());
            if (v.is_string())          return MCPToolHelpers::fromUtf8(v.get<std::string>());
            // Array/object — serialize to JSON text and let the adapter
            // store it as-is (useful for Postgres json/jsonb columns).
            return MCPToolHelpers::fromUtf8(v.dump());
        }

        std::string clientNameFromContext(const MCPToolContext& ctx)
        {
            return ctx.client.name.empty() ? std::string{"AI client"} : ctx.client.name;
        }
    }

    MCPToolResult InsertRowsTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);
        auto ts = MCPIdentifierValidator::extractTableAndSchema(params);

        if (!params.contains("rows") || !params["rows"].is_array() || params["rows"].empty())
            throw MCPToolError::invalidParameters("rows must be a non-empty array");

        const auto& rowsJson = params["rows"];

        // Permission + approval (Tier 3 always requires approval in RW mode).
        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        if (perm.requiresUserApproval())
        {
            // Preview first 3 rows for the dialog body.
            std::wstring preview;
            const int previewCap = (std::min<int>)(3, static_cast<int>(rowsJson.size()));
            for (int i = 0; i < previewCap; ++i)
            {
                preview += L"  " + MCPToolHelpers::fromUtf8(rowsJson[i].dump()) + L"\n";
            }
            if (static_cast<int>(rowsJson.size()) > previewCap)
                preview += L"  … and " +
                    std::to_wstring(rowsJson.size() - previewCap) + L" more\n";

            const std::wstring tableW = MCPToolHelpers::fromUtf8(ts.table);
            const std::wstring desc =
                L"Insert " + std::to_wstring(rowsJson.size()) +
                L" row(s) into '" + tableW + L"'";
            const std::wstring details = L"Preview:\n" + preview;

            const auto clientW = MCPToolHelpers::fromUtf8(clientNameFromContext(ctx));
            const bool ok = ctx.requestApproval(
                MCPToolHelpers::fromUtf8(this->name()),
                desc, details, connId, clientW);
            if (!ok) throw MCPToolError::permissionDenied("User denied the operation.");
        }

        auto [adapter, config] = ctx.getAdapter(connId);
        const std::wstring tableW = MCPToolHelpers::fromUtf8(ts.table);
        const std::wstring schemaW = ts.schema.has_value()
            ? MCPToolHelpers::fromUtf8(*ts.schema) : std::wstring{};

        int inserted = 0;
        for (const auto& rowJson : rowsJson)
        {
            if (!rowJson.is_object()) continue;
            TableRow row;
            for (auto it = rowJson.begin(); it != rowJson.end(); ++it)
            {
                const std::wstring key = MCPToolHelpers::fromUtf8(it.key());
                row[key] = jsonToCell(it.value());
            }
            try
            {
                auto r = adapter->insertRow(tableW, schemaW, row);
                if (r.success) ++inserted;
                else
                    throw MCPToolError::queryFailed(MCPToolHelpers::toUtf8(r.error));
            }
            catch (const std::exception& e)
            {
                throw MCPToolError::queryFailed(e.what());
            }
        }

        return MCPToolResult::text(
            "Inserted " + std::to_string(inserted) + " row(s) into '" + ts.table + "'.");
    }
}
